crates/sui-adapter/src/programmable_transactions/execution.rs

```diff
@@ -445,17 +449,96 @@ fn execute_move_publish<E: fmt::Debug, S: StorageView<E>, Mode: ExecutionMode>(
 /// Upgrade a Move package.  Returns an `UpgradeReceipt` for the upgraded package on success.
 fn execute_move_upgrade<E: fmt::Debug, S: StorageView<E>, Mode: ExecutionMode>(
     context: &mut ExecutionContext<E, S>,
-    _upgrade_ticket_arg: Argument,
-    _dep_ids: Vec<ObjectID>,
-    _module_bytes: Vec<Vec<u8>>,
+    upgrade_ticket_arg: Argument,
+    dep_ids: Vec<ObjectID>,
+    module_bytes: Vec<Vec<u8>>,
 ) -> Result<Vec<Value>, ExecutionError> {
     // Check that package upgrades are supported.
     context
         .protocol_config
         .check_package_upgrades_supported()
         .map_err(|_| ExecutionError::from_kind(ExecutionErrorKind::FeatureNotYetSupported))?;
 
-    invariant_violation!("Package upgrades are turned off. We should NEVER get here.")
+    assert_invariant!(
+        !module_bytes.is_empty(),
+        "empty package is checked in transaction input checker"
+    );
+
+    let upgrade_ticket: UpgradeTicket = {
+        let mut ticket_bytes = Vec::new();
+        let ticket_val: Value =
+            context.by_value_arg(CommandKind::Upgrade, 0, upgrade_ticket_arg)?;
+        let ticket_type = context
+            .session
+            .load_type(&TypeTag::Struct(Box::new(UpgradeTicket::type_())))
+            .map_err(|e| context.convert_vm_error(e))?;
+        check_param_type::<_, _, Mode>(context, 0, &ticket_val, &ticket_type)?;
+        ticket_val.write_bcs_bytes(&mut ticket_bytes);
+        bcs::from_bytes(&ticket_bytes).map_err(|_| {
+            ExecutionError::from_kind(ExecutionErrorKind::CommandArgumentError {
+                arg_idx: 0,
+                kind: CommandArgumentError::InvalidBCSBytes,
+            })
+        })?
+    };
+
+    // Run the move + sui verifier on the modules and publish them into the cache.
+    let upgraded_package_modules = publish_and_verify_modules::<_, _, Mode>(context, module_bytes)?;
+
+    // Check that this package ID points to a package and get the package we're upgrading.
+    let current_package = fetch_package(context, &upgrade_ticket.package.bytes)?;
+
+    // Read the package dependencies.
+    let dependency_packages = dep_ids
+        .iter()
+        .map(|id| fetch_package(context, id))
+        .collect::<Result<Vec<_>, _>>()?;
+
+    // TODO(tzakian): check digest of these packages against the digest that we expect in the
```

Comment:
This can be done first of all -- the digest is purely based on the module bytes and deps, after all.

Re: duplicates in the dependency list, yes, we should check that during linkage table verification, but that will be after this step, IIUC, and the main purpose is to prevent two versions of the same package from showing up, which we can only figure out once we have fetched the packages.

At this stage, it will always be safe to de-duplicate dep ids and then fetch them.